controllers/tlspolicy_status_updater.go

```diff
@@ -0,0 +1,233 @@
+	}
+}
+
+func (t *TLSPolicyStatusUpdaterReconciler) UpdateStatus(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, s *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("TLSPolicyStatusUpdaterReconciler").WithName("Reconcile")
+
+	policies := lo.FilterMap(topology.Policies().Items(), func(item machinery.Policy, index int) (*kuadrantv1alpha1.TLSPolicy, bool) {
+		p, ok := item.(*kuadrantv1alpha1.TLSPolicy)
+		return p, ok
+	})
+
+	for _, policy := range policies {
+		if policy.DeletionTimestamp != nil {
+			logger.V(1).Info("tls policy is marked for deletion, skipping", "name", policy.Name, "namespace", policy.Namespace)
+			continue
+		}
+
+		newStatus := &kuadrantv1alpha1.TLSPolicyStatus{
+			// Copy initial conditions. Otherwise, status will always be updated
+			Conditions:         slices.Clone(policy.Status.Conditions),
+			ObservedGeneration: policy.Status.ObservedGeneration,
+		}
+
+		var err error
+		accepted, ok := s.Load(TLSPolicyAcceptedKey(policy.GetUID()))
+		if ok && accepted != nil {
+			err = accepted.(error)
+		}
+		meta.SetStatusCondition(&newStatus.Conditions, *kuadrant.AcceptedCondition(policy, err))
+
+		// Do not set enforced condition if Accepted condition is false
+		if meta.IsStatusConditionFalse(newStatus.Conditions, string(gatewayapiv1alpha2.PolicyReasonAccepted)) {
+			meta.RemoveStatusCondition(&newStatus.Conditions, string(kuadrant.PolicyConditionEnforced))
+		} else {
+			enforcedCond := t.enforcedCondition(ctx, policy, topology)
+			meta.SetStatusCondition(&newStatus.Conditions, *enforcedCond)
+		}
+
+		// Nothing to do
+		equalStatus := equality.Semantic.DeepEqual(newStatus, policy.Status)
+		if equalStatus && policy.Generation == policy.Status.ObservedGeneration {
+			logger.V(1).Info("policy status unchanged, skipping update")
+			continue
+		}
+		newStatus.ObservedGeneration = policy.Generation
+		policy.Status = *newStatus
+
+		resource := t.Client.Resource(kuadrantv1alpha1.TLSPoliciesResource).Namespace(policy.GetNamespace())
+		un, err := controller.Destruct(policy)
+		if err != nil {
+			logger.Error(err, "unable to destruct policy")
+			continue
+		}
+
+		_, err = resource.UpdateStatus(ctx, un, metav1.UpdateOptions{})
+		if err != nil {
+			logger.Error(err, "unable to update status for TLSPolicy", "name", policy.GetUID(), "namespace", policy.GetNamespace(), "uid", policy.GetUID())
+		}
+	}
+
+	return nil
+}
+
+func (t *TLSPolicyStatusUpdaterReconciler) enforcedCondition(ctx context.Context, tlsPolicy *kuadrantv1alpha1.TLSPolicy, topology *machinery.Topology) *metav1.Condition {
+	if err := t.isIssuerReady(ctx, tlsPolicy, topology); err != nil {
+		return kuadrant.EnforcedCondition(tlsPolicy, kuadrant.NewErrUnknown(tlsPolicy.Kind(), err), false)
+	}
+
+	if err := t.isCertificatesReady(ctx, tlsPolicy, topology); err != nil {
+		return kuadrant.EnforcedCondition(tlsPolicy, kuadrant.NewErrUnknown(tlsPolicy.Kind(), err), false)
+	}
+
```

Comment:
TLS is an interesting case. If the gateway is not Programmed, it doesn't mean we cannot provision the certificates for its listeners. Is that why you didn't include the status of the gateway here?

Reply:
Exactly, a gateway can transition to a `Programmed` state from the certs created by the current tls policy controller.